### ISSUE TITLE
Source zip file posted on PyPi isn't source build

### ISSUE THREAD

**Issue by alerque** [robotools/defcon#354]
At least for 0.8.1 (and possibly others earlier, I didn't check) the source package posted on PyPi is not a proper source build. It seems to be a raw ZIP generated using `git archive` (possible from the GitHub source download links?). This is a problem because using the raw source archive or a Git clone requires the python setuptools SCM variant. The whole point of this SCM tooling is to make the release process earlier by not having to track some things in version control and let it figure out versioning from Git tags or whatever. But the Python build process is supposed to geterate a source release that is then compatible with regular python setuptools versions. SCM tooling should not be a requirement to build from the PyPi source downloads.

For all future releases please post source packages that are actually the **generated** source builds, not just archives of the repository. If you want help setting this up automated in CI feel free to poke me.

**Comment by anthrotype** [comment]
the sdist is built as usual `python setup.py sdist` and uploaded automatically here:
https://github.com/robotools/defcon/blob/6cfd94f7c649fdc72101de140cc28a61a2472291/.github/workflows/publish-package.yml#L42-L43

I don't know why you say "is not a proper source build. It seems to be a raw ZIP generated using git archive"

Can you describe the problem a bit in more detail? What error message do you get?

Maybe something has changed in upstream setuptools_scm that makes something downstream no longer work as expected.

**Comment by typemytype** [comment]
FYI: defcon changed to publishing on pypi with github actions see the PR https://github.com/robotools/defcon/pull/339
(just in case you are checking older version on pypi)

**Comment by anthrotype** [comment]
I see PR 339 added the `pyproject.toml` file, which supposedly enables a "build-system independent format for source trees" (https://www.python.org/dev/peps/pep-0517 and https://www.python.org/dev/peps/pep-0517).
Maybe your packaging tool does not support PEP517/518. You can tell pip to --no-use-pep517 and it will continue to run the "legacy" implicit setuptools based `setup.py install`, basically ignoring the pyproject.toml file. Try that and see if it fixes your problem.
But this is definitely _not_ an bug in defcon setup. I can install it from the .zip source distribution just fine in a fresh virtual environment with `pip install --no-binary=defcon decon==0.8.1`

**Comment by alerque** [comment]
> Can you describe the problem a bit in more detail? What error message do you get?

Building using `python setup.py` in a clean chroot with only `setuptools` fails with it complaining about needing `setuptools_scm`. That's what I was going on.

**Comment by anthrotype** [comment]
what version of setuptools?

I just did that (running `python setup.py install`) from a fresh virtual environment, and it worked for me. Python 3.9.4, setuptools 56.1.0

**Comment by anthrotype** [comment]
btw, running `setup.py install` will use the deprecated `easy_install` module, everybody including setuptools maintainers recommend to use pip to install third-party modules. I don't know how linux distros' build tools work internally, but I think they should no longer be calling the old `setup.py install` when building python modules for redistribution.

**Comment by alerque** [comment]
```console
$ pacman -Qi python-setuptools
Name            : python-setuptools
Version         : 1:56.0.0-1
```

Note I am not very deeply involved in Python, but I have an idea Python's idea of a "virtual env" is not the same as what I mean by a clean chroot. Is it possible that your virtual env testing is using the env for *some* dependencies but for ones it doesn't find it either goes and gets them with `pip` or finds them on the system? I suspect `pip` is actualaly installing or finding setuptools_scm in your env behind your back.

**Comment by anthrotype** [comment]
yes, because setuptools_scm is listed in setup.py's `setup_requires` (that's also how the setuptools_scm docs recommend to do and we do the same in all the other related projects); that makes setuptools fetch it from PyPI automatically when doing setup.py install. Maybe your "clean chroot" (no idea what that is) blocks or prevents setuptools from doing that somehow. The setup requirements are downloaded as eggs in an invisible `.eggs/` directory.

I think you should just change your build setup and require setuptools_scm, because it is really required for building the project. It's not a bug, but a feature.
 
Note that `git` itself is _not_ required. Setuptools_scm will look at the PKG-INFO file that is embedded in the source distribution zip file, and take the version string from there when it's present (not from the scm metadata).

**Comment by anthrotype** [comment]
(clicked the wrong button, i didn't mean to close the issue like that, apologies)

**Comment by alerque** [comment]
> because setuptools_scm is listed in setup.py's `setup_requires` (that's also how the setuptools_scm docs recommend to do and we do the same in all the other related projects); that makes setuptools fetch it from PyPI automatically when doing setup.py install.

True that's what is supposed to be in the Git source version of `setup.py`, but that's not what should end up in a *sdist*. Building *sdist* sources should not require SCM tooling.

I'll find a project that does this right and compare.

**Comment by alerque** [comment]
The fault seems to be in `Lib/defcon/__init__.py` L11, this should not be linking directly into setuptools_scm, it should be using the `__version__` string that SCM tooling wrote to a file at `Lib/defcon/_version.py`.

**Comment by benkiel** [comment]
@anthrotype just pushed a fix to the above, I'll cut a release and see if that does what you need. But, yes, the recommended install is to use `pip` not `python setup.py`

**Comment by alerque** [comment]
> But, yes, the recommended install is to use `pip` not `python setup.py`

I understand, but what's recommended for end users installing on their own systems is not always what is required for setting up distribution packages. I happen to be doing the latter.

**Comment by anthrotype** [comment]
yes, but we are not doing anything out of the ordinary here. We're simply using setuptools_scm the way it is designed to be used.
I don't see how this current defcon setup is any different from other projects where we also use the same setup.

**Comment by anthrotype** [comment]
>  just pushed a fix to the above

@benkiel did you? What did you do exactly? I don't see anything

**Comment by anthrotype** [comment]
> letting the build phase fetch code.

we are _not_ doing that. The build requirments (as specified in pyproject.toml as well as in the legacy `setup_requires` in setup.py) require that setuptools and setuptools_scm packages are present in order to build the package. If using a pep517 aware installer (like pip) these requirements are honuored at build time (pip creates a fresh env where it bootstraps the requirements and goes on building a wheel and installing off that). Otherwise, if installing with the `setup.py install` command, setuptools will honour the setup_requires (which include setuptools_scm).

As I have already said, setuptools_scm is really required for building the package from source, not just for building it from the git repository, but also when building off the sdist zip file. The reason it is required is because we don't want to hard-code the version string in the code itself, we want it to only be encoded in the package metadata (the standard `PKG-INFO` file). We also generate a `_version.py` module so that one can check `defcon.__version__` without needing to have pkg-resources or importlib.resources to be able to read the version from the metadata files (also the metadata files sometimes are not present when the module is embedded in applications).
Again, `git` is not required, only setuptools_scm is.

Whether you let pip or setuptools fetch the built-time dependencies, or your own packaging tool is responsible for that, it doesn't change that setuptools_scm is required to build (but not to run) the defcon module.

There is nothing wrong with having a build-time requirement. 

Is this a sufficient explanation?

**Comment by anthrotype** [comment]
your distro tool should really be updated to support pep517/518, the fact that we have a setup.py script and use setuptools as our build tool doesn't mean that we should continue to support distutils-like installation methods forever. What if we decide to use flit or poetry or meson or whatever pep517-compliant build tool (which measn we no longer have any setup.py script), would that mean you can redistribute our package?

**Comment by alerque** [comment]
> We're simply using setuptools_scm the way it is designed to be used.
> I don't see how this current defcon setup is any different from other projects where we also use the same setup.

It is different. I can build distro packages for most Python projects in a chroot with just `setuptools`, including the ones that use `setuptools_scm` internally. There is something different here because the Git clone / archive requiring scm tools should not result in the sdist requiring scm tools. The sdist build process should do all the scm things then leave the package with just a regular setuptools dependency. Somehow this works for many other projects, I haven't found what's different here yet.

**Comment by anthrotype** [comment]
> It is different

please show me where it is, I can't see how it is different

> sdist requiring scm tools

**Comment by anthrotype** [comment]
> Somehow this works for many other projects, I haven't found what's different here yet

that's strange, and I don't have an answer yet for your I'm afraid.

> sdist requiring scm tools

it does not require "scm" tools as in the git command. It requires setuptools_scm, only for building the package, to extract version string from the PKG-INFO metadata file embedded in the source distribution zip package, as I have explained above.

> leave the package with just a regular setuptools dependency

that's your tool's limit. python packaging is not restricted to using setuptools only nowadays.

> Dude

sorry, what did I say that make you "dude" me? I'm honestly trying to help understanding the cause of the issue.

> Python's setuptools_scm should take care of all the scm bits during sdist generation and it shouldn't be required afterwards.

that's not how setuptools_scm works, I'm afraid. You're free to raise the issue with the setuptools_scm maintainers if you like.

**Comment by madig** [comment]
Uh, there's a `setup_requires` entyr in setup.cfg and setup.py and a build system entry in pyproject.toml. Maybe try cleaning that up and leaving only the pyproject.toml as the source of setuptools_scm?

**Comment by benkiel** [comment]
> @benkiel did you? What did you do exactly? I don't see anything

Sorry, I saw that you fixed the .gitignore for the _version_.py, that's what I meant

**Comment by anthrotype** [comment]
i don't think that's related, the `_version.py` file was already included in the source distribution anyway, the gitignore change was only to make sure you don't accidentally commit the file

**Comment by alerque** [comment]
I did some research ... and ended up down a rabbit hole without a bottom. Apparently several things I thought I *knew* are no longer true in more recent setuptools_scm versions. I'm not the only one with complaints and I found lots of distro package managers with similar complaints, but I'm convinced it is out of scope of this project to fix. The general controversy over Python build systems is galactic in scale and I'm not going to wade in deeper right now. I'll just borrow some hacks around the problem(s) for now. Unfortunately pep0517 is a step backwards for distro packages so I appreciate setuptools hasn't been dropped yet, but that's also not really this project's concern.

In goot news, I did discover there is a way to get version information for tagged commits out of a `git archive`. In spite of my Git ninja foo this was news to me. Not needed for this project per se but see [`setuptools-scm-git-archive`](https://pypi.org/project/setuptools-scm-git-archive/) if you run a Python project, don't want to bother with `sdist` at all, but also don't want to hard code the versions.

I'm off to write some Cargo & GNU Autotool macros to do the same.
